cargo-dist/src/tasks.rs

```diff
@@ -1103,6 +1108,134 @@ impl<'pkg_graph> DistGraphBuilder<'pkg_graph> {
+
+        // If they have an x64 macos build but not an arm64 one, add a fallback entry
+        // to try to install x64 on arm64 and let rosetta2 deal with it.
+        //
+        // (This isn't strictly correct because rosetta2 isn't installed by default
+        // on macos, and the auto-installer only triggers for "real" apps, and not CLIs.
+        // Still, we think this is better than not trying at all.)
+        const X64_MACOS: &str = "x86_64-apple-darwin";
+        const ARM64_MACOS: &str = "aarch64-apple-darwin";
+        let mut has_x64_apple = false;
+        let mut has_arm_apple = false;
+        for &variant_idx in &release.variants {
+            let variant = self.variant(variant_idx);
+            let target = &variant.target;
+            if target == X64_MACOS {
+                has_x64_apple = true;
+            }
+            if target == ARM64_MACOS {
+                has_arm_apple = true;
+            }
+        }
+        let do_rosetta_fallback = has_x64_apple && !has_arm_apple;
+
+        let mut arm64 = None;
+        let mut x86_64 = None;
+
+        // Gather up the bundles the installer supports
+        let mut artifacts = vec![];
+        let mut target_triples = SortedSet::new();
+        for &variant_idx in &release.variants {
+            let variant = self.variant(variant_idx);
+            let target = &variant.target;
+            if target.contains("windows") || target.contains("linux-gnu") {
+                continue;
+            }
+            // Compute the artifact zip this variant *would* make *if* it were built
+            // FIXME: this is a kind of hacky workaround for the fact that we don't have a good
+            // way to add artifacts to the graph and then say "ok but don't build it".
+            let (artifact, binaries) =
+                self.make_executable_zip_for_variant(to_release, variant_idx);
+            target_triples.insert(target.clone());
+            let fragment = ExecutableZipFragment {
+                id: artifact.id,
+                target_triples: artifact.target_triples,
+                zip_style: artifact.archive.as_ref().unwrap().zip_style,
+                binaries: binaries
+                    .into_iter()
+                    .map(|(_, dest_path)| dest_path.file_name().unwrap().to_owned())
+                    .collect(),
+            };
+
+            if target == X64_MACOS {
+                x86_64 = Some(fragment.clone());
+            }
+            if target == ARM64_MACOS {
+                arm64 = Some(fragment.clone());
+            }
+
+            if do_rosetta_fallback && target == X64_MACOS {
+                // Copy the info but respecify it to be arm64 macos
+                let mut arm_fragment = fragment.clone();
+                arm_fragment.target_triples = vec![ARM64_MACOS.to_owned()];
+                artifacts.push(arm_fragment.clone());
+                arm64 = Some(arm_fragment);
+            }
+            artifacts.push(fragment);
```

Comment:
I assume it's intentional that even though we're doing fallback we're not actually assigning the x64 binary to the `arm64` var because we want homebrew et al to understand what's going on?